detectors/gcp/gce.go

```diff
@@ -28,30 +28,41 @@ import (
 )
 
 // GCE collects resource information of GCE computing instances
-type GCE struct{}
+type GCE struct {
+	mc    metadataClient
+	onGCE func() bool
+}
+
+var NewGCE = func() *GCE {
+	return &GCE{
+		mc:    metadata.NewClient(nil),
+		onGCE: metadata.OnGCE,
+	}
+}
 
 // compile time assertion that GCE implements the resource.Detector interface.
 var _ resource.Detector = (*GCE)(nil)
 
 // Detect detects associated resources when running on GCE hosts.
 func (gce *GCE) Detect(ctx context.Context) (*resource.Resource, error) {
-	if !metadata.OnGCE() {
+	if !gce.onGCE() {
```

Comment:
This is an exported type in a stable package. It needs to be considered what will happen for an existing user that is declaring a new variable of this type and they upgrade: the `onGCE` field will be nil and this will panic.

This also exports a new variable (I'm not sure why is this isn't `func` declaration?). I would suggest we try and preserve the existing initialization behavior of the `GCE` type as best we can to support existing uses and not include this new `var`. Also, if we ultimately want to add this, it should not be done in a PR dedicated to the addition (not in a PR adding semconv values).

I expect this was done to accommodate mocking while testing. Instead of this approach, I would suggest abstract the `metadata` package functions used to be unexported package variables that a test could overwrite.